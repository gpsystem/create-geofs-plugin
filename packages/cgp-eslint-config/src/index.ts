import merge from "lodash.merge";
import {
  eslintConfigBaseDependencies,
  eslintConfigBases,
  EslintConfigNames,
  UnknownEslintConfigBase,
} from "./configurations";

export type MoreThanOneArray<T> = [T, ...T[]];

// TODO: is there a type for a eslint configuration file?
/**
 * Merges eslint configuration templates into one configuration to output to a .eslintrc file.
 *
 * @param configNames All the configuration templates.
 * The templates will be merged (deeply) with further right configuration templates overwriting further left templates.
 *
 * @returns A tuple. First is the configuration, ready to be stringified and outputted to a .eslintrc file.
 * Second is the names of the dependencies to install to make sure the .eslintrc configuration works properly.
 */
export function getEslintConfig(
  ...configNames: MoreThanOneArray<EslintConfigNames>
): [config: Record<string, unknown>, dependencies: string[]] {
  if (!configNames.length)
    throw new Error("must provide at least one config name");
  const configsToMerge: UnknownEslintConfigBase[] = configNames.map(
    (val) => eslintConfigBases[val]
  );
  const config: Record<string, unknown> = merge(
    ...(configsToMerge as MoreThanOneArray<UnknownEslintConfigBase>)
  );

  // if the typescript template is mentioned, remove globals from the config
  if (configNames.includes("tsBase") && config.globals !== undefined)
    delete config.globals;

  return [
    config,
    configNames.reduce<string[]>(
      (acc, val) => [...acc, ...eslintConfigBaseDependencies[val]],
      []
    ),
  ];
}

export { EslintConfigNames };
