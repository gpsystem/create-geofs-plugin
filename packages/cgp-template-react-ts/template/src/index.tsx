/// <reference types="jquery" />
/// <reference types="@geps/geofs-types" />

import React from "react";
import { createRoot } from "react-dom/client";
import { App } from "./App";

// Replace the selector with a link to your UI host!
// Example: #geofs-ui-right
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const root = createRoot(document.querySelector("REPLACE ME")!);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
