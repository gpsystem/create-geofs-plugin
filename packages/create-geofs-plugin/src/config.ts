import { resolve } from "node:path";
import { cwd } from "node:process";
import { Command } from "commander";
import { sync as packageDirectorySync } from "pkg-dir";
import { version } from "../package.json";

export interface Configuration {
  targetDir: string;
  templateName: "default" | "typescript" | "react" | "react-ts";
  templateDir: string;
  overwrite: boolean;
}

export function parseConfig(argv: string[]): Configuration {
  let destination = "";

  // eslint-disable-next-line prefer-const
  let { overwrite, template } = new Command()
    .name("create-geofs-plugin")
    .version(`Create GeoFS Plugin v${version}`)
    .argument("<destination>")
    .action((dest: string) => {
      if (!dest.includes("/"))
        throw new Error("argument <destination> must be a relative path");
      else destination = resolve(cwd(), dest);
    })
    .option("--overwrite", "Overwrite existing files", false)
    .option("-t, --template <template>", "Name of the template to use")
    .parse(argv)
    .opts<{ overwrite?: boolean; template?: string }>();

  if (!["default", "typescript", "react", "react-ts"].includes(template ?? ""))
    template = "default";

  // TODO: create templates
  const templateDir: string | undefined = require.resolve(
    // `@geps/cgp-template-${template}`
    `@geps/cgp-eslint-config`
  );
  if (!templateDir) throw new Error("template not found");

  return {
    targetDir: destination,
    overwrite: overwrite ?? false,
    templateName: template as "default" | "typescript" | "react" | "react-ts",
    templateDir,
  };
}
