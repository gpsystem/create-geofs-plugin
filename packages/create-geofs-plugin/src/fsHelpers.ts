import {
  Dirent,
  copyFileSync,
  existsSync,
  mkdirSync,
  readdirSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";

function pathIsDirectory(path: string): boolean {
  return statSync(path).isDirectory();
}

function internalWriteFileSync(targetFile: string, data: string): void {
  writeFileSync(targetFile, data, "utf-8");
}

export function outputFile(
  targetFile: string,
  data: string,
  force = false
): void {
  if (!force && existsSync(targetFile)) throw new Error("file already exists");

  const targetFileDirectory: string = dirname(targetFile);

  if (!existsSync(targetFileDirectory))
    mkdirSync(targetFileDirectory, { recursive: true });
  internalWriteFileSync(targetFile, data);
}

// TODO: should we make a wrapper for this that throws on initialization and returns the full string[]?
export function* getAllFilesInDir(
  dirName: string
): Generator<string, void, void> {
  if (!pathIsDirectory(dirName))
    throw new Error(`passed path is not a directory`);

  const directoryDetails: Dirent[] = readdirSync(dirName, {
    withFileTypes: true,
  });

  for (const directoryChild of directoryDetails) {
    const childPath: string = resolve(dirName, directoryChild.name);
    const isDirectory = directoryChild.isDirectory();

    if (isDirectory) yield* getAllFilesInDir(childPath);
    else yield childPath;
  }
}

export function copyDir(
  srcDir: string,
  targetDir: string,
  overwrite = false
): void {
  if (pathIsDirectory(srcDir)) throw new Error("source path isn't a directory");

  for (const srcFilePath of getAllFilesInDir(srcDir)) {
    const newFilePath: string = join(targetDir, relative(srcDir, srcFilePath));

    mkdirSync(dirname(newFilePath), { recursive: true });
    copyFileSync(srcFilePath, newFilePath);
  }
}
