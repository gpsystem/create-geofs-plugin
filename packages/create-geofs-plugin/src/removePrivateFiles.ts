export default function removePrivateFiles(directoryPath: string) {
  // all private files start with __
}
