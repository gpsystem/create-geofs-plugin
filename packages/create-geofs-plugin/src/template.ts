import { readFileSync } from "node:fs";
import { EslintConfigNames, getEslintConfig } from "@geps/cgp-eslint-config";

export interface TemplateJson {
  /** The dependencies of the template. Does not include the dependencies returned from cgp-eslint-config. */
  dependencies: string[];
  eslintConfigTemplates: EslintConfigNames[];
}

function isObject(obj: unknown): obj is Record<string, unknown> {
  // intentionally allowing a comparison of both undefined and null
  // Object.getPrototypeOf will throw for both undefined and null
  // eslint-disable-next-line eqeqeq
  return obj != null && Object.getPrototypeOf(obj) === Object.prototype;
}

export function assertTemplateJsonFormat(
  templateJson: unknown
): templateJson is TemplateJson {
  // thanks to Joshua Chen on the TS discord for this assertion code!
  type AssertTrue<T extends true> = T;
  // If typescript throws type 'false' does not satisfy the constraint 'true' here, then
  // allPossibleConfigNames is malformed
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  type A = AssertTrue<
    [EslintConfigNames] extends [typeof allPossibleConfigNames[number]]
      ? true
      : false
  >;

  const allPossibleConfigNames = ["baseConfig", "tsBase", "reactBase"] as const;
  const { isArray } = Array;

  return (
    isObject(templateJson) &&
    isArray(templateJson.dependencies) &&
    templateJson.dependencies.every<string>((val): val is string => {
      // TODO: ensure val is a npm dep name followed by a major version number (after @ symbol)
      return typeof val === "string";
    }) &&
    isArray(templateJson.eslintConfigTemplates) &&
    templateJson.eslintConfigTemplates.every<EslintConfigNames>(
      (val): val is EslintConfigNames => allPossibleConfigNames.includes(val)
    )
  );
}

export interface ExpandedTemplateJson {
  /** All dependencies of the template, including the eslint dependencies. */
  dependencies: string[];
  /** The configuration to save to .eslintrc.yml. */
  eslintConfig: Record<string, unknown>;
}

export default function expandTemplateJson(
  templateJsonPath: string
): ExpandedTemplateJson {
  const templateJson: unknown = JSON.parse(
    readFileSync(templateJsonPath, "utf-8")
  );

  if (!assertTemplateJsonFormat(templateJson))
    throw new Error("template.json is malformed");

  const { eslintConfigTemplates } = templateJson;
  if (eslintConfigTemplates.length === 0)
    throw new Error("template must specify at least one config template");
  if (eslintConfigTemplates[0] !== "baseConfig")
    throw new Error(
      "template's eslint config template list must include the base configuration first"
    );

  const [eslintConfig, eslintDependencies] = getEslintConfig(
    ...(eslintConfigTemplates as [EslintConfigNames, ...EslintConfigNames[]])
  );

  const dependencies: string[] = [
    ...templateJson.dependencies,
    ...eslintDependencies,
  ];

  return {
    dependencies,
    // TODO
    eslintConfig: eslintConfig as Record<string, unknown>,
  };
}
