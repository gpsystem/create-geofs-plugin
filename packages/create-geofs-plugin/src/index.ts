import { join } from "node:path";
import { dir, DirectoryResult } from "tmp-promise";
import { Configuration, parseConfig } from "./config";
import { copyDir } from "./fsHelpers";
import expandTemplateJson, { ExpandedTemplateJson } from "./template";

export async function start(argv: string[]): Promise<void> {
  const config: Configuration = parseConfig(argv);
  const { path: tmpDirName, cleanup: cleanupTempDir }: DirectoryResult =
    await dir({
      unsafeCleanup: true,
    });

  copyDir(config.templateDir, tmpDirName, { overwrite: config.overwrite });
  const { dependencies: templateDeps, eslintConfig }: ExpandedTemplateJson =
    expandTemplateJson(join(tmpDirName, "template.json"));
  // remove everything that starts with __
  // copy everything to the target dir
  // install the dependencies
  await cleanupTempDir();
}
