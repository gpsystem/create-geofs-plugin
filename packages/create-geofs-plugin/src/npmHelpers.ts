import { execSync } from "node:child_process";
import { join } from "node:path";
import { outputFile } from "./fsHelpers";

export function initializeNpm(
  targetDir: string,
  templateDeps: string[],
  ci: boolean
): void {
  const firstPackageJson: Record<string, unknown> = {
    name: "geofs-plugin",
    version: "",
    private: true,
  };

  outputFile(
    join(targetDir, "package.json"),
    JSON.stringify(firstPackageJson, null, 2),
    {
      force: true,
    }
  );
  if (!ci) execSync(`npm i ${templateDeps.join(" ")} -D`, { cwd: targetDir });
}
