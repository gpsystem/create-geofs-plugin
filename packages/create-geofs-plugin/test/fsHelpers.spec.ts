import {
  existsSync,
  mkdirSync,
  readFileSync,
  rmSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, sep } from "node:path";
import { copyDir, getAllFilesInDir, outputFile } from "../src/fsHelpers";
import { testTargetDir } from "./utils/index";

function createTargetDir(): void {
  mkdirSync(testTargetDir);
}

function deleteTargetDir(): void {
  rmSync(testTargetDir, { recursive: true, force: true });
}

function joinWithTargetDir(...paths: string[]): string {
  return join(testTargetDir, ...paths);
}

describe("output file", () => {
  beforeEach(createTargetDir);
  afterEach(deleteTargetDir);

  test("throws on already existing file", () => {
    const filePath: string = joinWithTargetDir("first", "test.txt");

    mkdirSync(dirname(filePath));
    writeFileSync(filePath, "1");

    expect(() => {
      outputFile(filePath, "2");
    }).toThrowErrorMatchingSnapshot();
  });

  test("overwrites file when force=true", () => {
    const filePath: string = joinWithTargetDir("first", "test.txt");

    mkdirSync(dirname(filePath));
    writeFileSync(filePath, "1");

    const newFileData = "2";
    outputFile(filePath, newFileData, true);

    expect(readFileSync(filePath, "utf-8")).toBe(newFileData);
  });

  test("creates directories when they don't already exist", () => {
    const directoryPath: string = joinWithTargetDir("first", "second");
    const filePath: string = join(directoryPath, "test.txt");
    const fileContents = "1";

    outputFile(filePath, fileContents);

    expect(existsSync(directoryPath)).toBe(true);
  });

  test("doesn't create an already existing directory", () => {
    const directoryPath: string = joinWithTargetDir("first/");
    const firstFilePath: string = join(directoryPath, "test.txt");
    const firstFileContents = "1";

    mkdirSync(directoryPath);
    writeFileSync(firstFilePath, firstFileContents);
    outputFile(join(directoryPath, "test2.txt"), "2");

    // if it recreated the directory, the first file would no longer exist
    expect(existsSync(firstFilePath)).toBe(true);
  });

  test("writes correctly", () => {
    // DO NOT TOUCH THE CONSTANTS
    // if you do touch these constants, remember to regenerate the snapshots
    const filePath: string = joinWithTargetDir("first", "file.txt");
    const fileContents = "test contents, 12345, 1.2345";

    outputFile(filePath, fileContents);

    expect(readFileSync(filePath, "utf-8")).toMatchSnapshot(
      relative(testTargetDir, filePath).split(sep).join("/")
    );
  });
});

describe("get all files in a directory", () => {
  // use only relative paths from testTargetDir
  const allFiles: string[] = [
    "./file.txt",
    "./.hidden/file.txt",
    "./some/deeply/nested/file.txt",
  ];

  beforeEach(() => {
    createTargetDir();
    for (const fileName of allFiles) {
      outputFile(joinWithTargetDir(fileName), "test data");
    }
  });
  afterEach(deleteTargetDir);

  test("throws when passed path isn't a directory", () => {
    expect(() => [
      ...getAllFilesInDir(joinWithTargetDir("file.txt")),
    ]).toThrowErrorMatchingSnapshot();
  });

  test("always returns real paths", () => {
    const allPaths: string[] = [...getAllFilesInDir(testTargetDir)];

    for (const filePath of allPaths) {
      expect(statSync(filePath).isFile()).toBe(true);
    }
  });

  test("gets the correct paths", () => {
    const allPaths: string[] = [...getAllFilesInDir(testTargetDir)];

    expect(
      allFiles.map((relPath) => joinWithTargetDir(relPath)).sort()
    ).toEqual(allPaths.sort());
  });
});

describe("copy directories", () => {
  test.todo("throws when the passed source path isn't a directory");

  test.todo("skips files that already exist");

  test.todo("overwrites files when overwrite=true");

  test.todo("doesn't modify directory structure");

  test.todo("doesn't modify file contents");
});
