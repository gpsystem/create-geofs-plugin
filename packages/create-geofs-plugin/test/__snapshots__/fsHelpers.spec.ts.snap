// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get all files in a directory throws when passed path isn't a directory 1`] = `"passed path is not a directory"`;

exports[`output file throws on already existing file 1`] = `"file already exists"`;

exports[`output file writes correctly: first/file.txt 1`] = `"test contents, 12345, 1.2345"`;
